smtp.go

```diff
@@ -2,209 +2,334 @@ package maddy
+		r, cont, err := step.Pass(&ctx, currentMsg)
+		if !cont {
+			if err == module.ErrSilentDrop {
+				return nil
+			}
+			if err != nil {
+				return err
+			}
 		}
-	}
 
-	if tokens, ok := tokens["pgp"]; ok {
-		var err error
-		be, err = newSMTPPGP(caddyfile.NewDispenserTokens("", tokens), be)
-		if err != nil {
-			return nil, err
+		if r != nil {
+			if _, err := io.Copy(&buf, r); err != nil {
+				return err
+			}
+			currentMsg = &buf
 		}
 	}
-
-	s := smtp.NewServer(newVerifier(newRelay(be)))
-	s.LMTP = lmtp
-	s.Domain = hostname
-	return s, nil
+	return nil
 }
 
-type newUpstreamFunc func(d caddyfile.Dispenser) (smtp.Backend, error)
-
-func newUpstream(tokens map[string][]caddyfile.Token) (smtp.Backend, error) {
-	upstreams := map[string]newUpstreamFunc{
-		"proxy": newSMTPProxy,
-		"debug": newSMTPDebug,
-	}
-
-	var be smtp.Backend
-	for name, newUpstream := range upstreams {
-		tokens, ok := tokens[name]
-		if !ok {
-			continue
-		}
-
-		if be != nil {
-			return nil, errors.New("multiple SMTP upstreams specified")
-		}
+func (u SMTPUser) Logout() error {
+	return nil
+}
 
-		var err error
-		be, err = newUpstream(caddyfile.NewDispenserTokens("", tokens))
-		if err != nil {
-			return nil, fmt.Errorf("failed to create `%v` upstream: %v", name, err)
-		}
-	}
+type SMTPEndpoint struct {
+	Auth      module.AuthProvider
+	serv      *smtp.Server
+	name      string
+	domain    string
+	listeners []net.Listener
+	pipeline  []SMTPPipelineStep
+}
 
-	if be == nil {
-		return nil, errors.New("missing SMTP upstream")
-	}
+func (endp *SMTPEndpoint) Name() string {
+	return "smtp"
+}
 
-	return be, nil
+func (endp *SMTPEndpoint) InstanceName() string {
+	return endp.name
 }
 
-func newSMTPProxy(d caddyfile.Dispenser) (smtp.Backend, error) {
-	for d.Next() {
-		args := d.RemainingArgs()
+func (endp *SMTPEndpoint) Version() string {
+	return VersionStr
+}
 
-		if len(args) == 1 {
-			addr, err := standardizeAddress(args[0])
+func NewSMTPEndpoint(instName string, cfg config.CfgTreeNode) (module.Module, error) {
+	endp := new(SMTPEndpoint)
+	endp.name = instName
+	var (
+		err     error
+		tlsConf *tls.Config
+		tlsSet  bool
+
+		localDeliveryDefault  string
+		localDeliveryOpts     map[string]string
+		remoteDeliveryDefault string
+		remoteDeliveryOpts    map[string]string
+	)
+	maxIdle := -1
+	maxMsgBytes := -1
+
+	insecureAuth := false
+	ioDebug := false
+	for _, entry := range cfg.Childrens {
+		switch entry.Name {
+		case "auth":
+			endp.Auth, err = authProvider(entry.Args)
+			if err != nil {
+				return nil, err
+			}
+		case "hostname":
+			if len(entry.Args) != 1 {
+				return nil, errors.New("hostname: expected 1 argument")
+			}
+			endp.domain = entry.Args[0]
+		case "maxidle":
+			if len(entry.Args) != 1 {
+				return nil, errors.New("maxidle: expected 1 argument")
+			}
+			maxIdle, err = strconv.Atoi(entry.Args[0])
+			if err != nil {
+				return nil, errors.New("maxidle: invalid integer value")
+			}
+		case "maxmsgbytes":
```

Comment:
Could use "size" instead, so we can add units later on (`max_message_size 1MB`).

Reply:
Ok. Also, found out that I forgot to include this option in CONFIG_REFERENCE.md.